### PR TITLE
Properly disable echoing in getPassword when running in MinTTY

### LINKED ISSUE #50: getPassword reveals your password as you type it on MSYS2 (Windows)
This same issue bit `cabal-install` (https://github.com/haskell/cabal/issues/4128), and it affects `haskeline` too. Not only is this Windows-specific, but it's also console-specific, as it requires the MSYS2 (or Cygwin) console (as opposed to `cmd.exe` or PowerShell, which are exempt from this issue). To reproduce, clone this repo in MSYS2 and run:

```
$ runghc haskeline/examples/Test.hs
0:password
line 0:password
1:hunter2
line 1:hunter2
2:q
```

Notice that it echoed my password (`hunter2`) when I typed it! This is because sadly, the `hGetEcho` and `hSetEcho` functions (which `getPassword` uses internally) don't work properly on MinTTY consoles like MSYS2. To disable echoing on MinTTY, you have to shell out to `stty`.

I fixed the `cabal-install` issue by creating my own [`echo`](http://hackage.haskell.org/package/echo) library, which implements all of the functionality needed to have a `withoutInputEcho` function that works properly on all operating systems and consoles, and made `cabal-install` depend on it. However, given that `haskeline` is a GHC dependency, that probably won't work in this case. Would you object if I inlined the code from `echo` into `haskeline` to fix `getPassword`?

### PR DESCRIPTION
**Pull request by RyanGlScott**
This copies over functionality from the `echo` library that, on Windows, detects if we're running in a MinTTY console, and if so, shells out to `stty` to disable/enable input echoing instead of `hSetEcho`/`hGetEcho`, which are broken on MinTTY.

Fixes #50.

### CODE REVIEW

File: System/Console/Haskeline/Backend/Win32/MinTTY.hsc [hunk 1]
```diff
@@ -0,0 +1,235 @@
+{-
```
Comment (judah) [inline System/Console/Haskeline/Backend/Win32/MinTTY.hsc]: Given that the module already exists in newer versions of Win32, I'd prefer to do something simpler:
- Only do this fix when Win32>=2.5
- Add a comment to getInputPassword describing the bug in the meantime

Otherwise, I'm concerned about the extra level of complexity we're adding in order to fix this corner case (and my ability to maintain it in the long run), especially since that there's already other ways Haskeline behaves suboptimally on MinTTY.

File: System/Console/Haskeline/Echo.hs [hunk 2]
```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE CPP #-}
+
+{-|
+Module:      System.Console.Haskeline.Echo
```
Comment (judah) [inline System/Console/Haskeline/Echo.hs]: I prefer to avoid these headers in internal-only modules.  They also seem out of place (for example the LICENCE is actually BSD).

File: System/Console/Haskeline.hs [hunk 3]
```diff
@@ -244,8 +245,7 @@ getPassword x = promptedInput
                                         $ runCommandLoop tops prefix loop
                                         $ Password [] x)
                     (\fops -> let h_in = inputHandle fops
-                              in bracketSet (hGetEcho h_in) (hSetEcho h_in) False
-                                  $ runMaybeT $ getLocaleLine fops)
+                              in withoutInputEcho h_in $ runMaybeT $ getLocaleLine fops)
```
Comment (judah) [inline System/Console/Haskeline.hs]: Were you able to make sense of my comment from: https://github.com/judah/haskeline/issues/50#issuecomment-271436812 about integrating this with FileOps and Backend?  I think it would help reduce complexity around the code splits between the Posix and Windows backends.

File: System/Console/Haskeline/Echo.hs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+echoOff = if minTTY then MinTTY "-echo" else DefaultTTY False
+
+-- | Settings used to configure the @stty@ command-line utility.
+type STTYSettings = String
+
+-- | Is the current process attached to a MinTTY console (e.g., Cygwin or MSYS)?
+minTTY :: Bool
+#if defined(MINGW)
+minTTY = unsafePerformIO $ do
+  h <- getStdHandle sTD_INPUT_HANDLE
```
Comment (RyanGlScott) [inline System/Console/Haskeline/Echo.hs]: OK. But keep in mind that Haskell `Handle`s and Win32 `HANDLE`s (the latter of which `isMinTTYHandle` accepts as an argument) are not the same thing, and you have to do a bit of an elaborate dance to convert from the former to the latter. This would require bringing in [some more code](https://github.com/haskell/win32/commit/729f9029e7a1fb8274c0b7f8065dec45f003405a#diff-0546d95d825d023b9c8315fd6043612eR253), which I imagine will be in `Win32-2.6`.

Given your comments [here](https://github.com/judah/haskeline/pull/52#discussion_r98361809), would you be OK with putting this much code into `haskeline`? Or do you want to wait until `Win32-2.6`'s release so that you can import both `isMinTTYHandle` _and_ `withHandleToHANDLE` from `Win32` directly?

File: System/Console/Haskeline.hs (hunk 3 above)
Comment (RyanGlScott) [inline System/Console/Haskeline.hs]: Honestly, no, that doesn't quite make sense to me. You say that this change would mean we could get rid of `inputHandle`, but that would obviate the need for https://github.com/judah/haskeline/pull/52#discussion_r98360675! So I'm a bit lost.

File: System/Console/Haskeline/Echo.hs (hunk 2 above)
Comment (RyanGlScott) [inline System/Console/Haskeline/Echo.hs]: Done.

File: System/Console/Haskeline/Backend/Win32/MinTTY.hsc (hunk 1 above)
Comment (RyanGlScott) [inline System/Console/Haskeline/Backend/Win32/MinTTY.hsc]: Done.

File: System/Console/Haskeline.hs (hunk 3 above)
Comment (judah) [inline System/Console/Haskeline.hs]: Ah, got it.  What I meant is that the inputHandle would still be there implicitly, but be encapsulated within the closure.  (Which is already happening for other fields in `FileOps`, e.g., `getLocaleLine`.)

In `System.Console.Haskeline.Term`:

    data RunTerm { ..., termOps = Either TermOps FileOps }
    data FileOps = FileOps {
            withoutInputEcho :: forall a . IO a -> IO a,  -- This is the new line
            wrapFileInput :: forall a . IO a -> IO a,  -- This is an existing field
            ...
        }

and in `System.Console.Haskeline.Backend.Win32`:
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend/Win32.hsc#L403

    fileRunTerm :: Handle -> IO RunTerm
    fileRunTerm h_in = do
        ...
        return RunTerm {
            ...,
            termOps = Right FileOps {
                wrapFileInput = hWithBinaryMode h_in,
                withoutInputEcho = hWithoutInputEcho h_in,
               ...
              }
         }
      
where `hWithoutInputEcho` is the function from your `Echo` module.  That way the `FileOps` implicitly captures whichever `Handle` was specified in the `Backend`.

Finally, replace the code from `getPassword` https://github.com/judah/haskeline/blob/master/System/Console/Haskeline.hs#L246
with:

    getPassword x = promptedInput (...)
                    (\fops ->  withoutInputEcho fops False
                                  $ runMaybeT $ getLocaleLine fops)

The value of fops (and thus, the Handle that it wraps) that gets piped in will depend on which backend is used (the logic is in System.Console.Haskeline.Backend).  For reference (you don't need to change anything), these are the places that pass in the handle to it:
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend/Win32.hsc#L382
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend.hs#L54

File: System/Console/Haskeline/Echo.hs (hunk 4 above)
Comment (judah) [inline System/Console/Haskeline/Echo.hs]: Hm.  In this case, `withHandletoHANDLE` is only about 10 lines of code (plus imports and comments), right?  In that case it seems simple enough to inline into Haskeline.

I guess neither `Win32-2.5` nor `Win32-2.6` has been released as part of `ghc-8.0`; do you know if it's likely that `ghc-8.2` will include `Win32-2.6`?  If so, then we might as will make Haskeline require `Win32-2.6`; but if not, then I think it's reasonable to copy that code into Haskeline and only require `Win32-2.5`.  (If you're not sure what will happen, then let's just inline it for now.)

Comment (RyanGlScott) [inline System/Console/Haskeline/Echo.hs]: GHC 8.2 will be released with `Win32-2.5`, so I'll just inline `withHandleToHANDLE`.

**Comment by RyanGlScott** [comment]
Thanks for your feedback, @judah. I've shuffled the code around so that `withoutInputEcho` is now a field of `FileOps`. I've also made `minTTY` take a `Handle` as an argument.

File: haskeline.cabal [hunk 5]
```diff
@@ -20,7 +20,7 @@ Homepage:       http://trac.haskell.org/haskeline
 Bug-Reports:    https://github.com/judah/haskeline/issues
 Stability:      Stable
 Build-Type:     Custom
-extra-source-files: examples/Test.hs Changelog
+extra-source-files: examples/Test.hs Changelog includes/*.h
```
Comment (judah) [inline haskeline.cabal]: Can you explain why this change was needed (vs how the existing header win_console.h is treated)?

File: System/Console/Haskeline/Backend/Win32/Echo.hs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+hEchoOff :: Handle -> EchoState
+hEchoOff input = if minTTY input then MinTTY "-echo" else DefaultTTY False
+
+-- | Settings used to configure the @stty@ command-line utility.
+type STTYSettings = String
+
+-- | Is the current process attached to a MinTTY console (e.g., Cygwin or MSYS)?
+minTTY :: Handle -> Bool
+#if MIN_VERSION_Win32(2,5,0)
+minTTY input = unsafePerformIO $ withHandleToHANDLE input isMinTTYHandle
```
Comment (judah) [inline System/Console/Haskeline/Backend/Win32/Echo.hs]: Can we make this `Handle -> IO Bool`, and remove the `unsafePerformIO` and `NOINLINE`?

Comment (RyanGlScott) [inline System/Console/Haskeline/Backend/Win32/Echo.hs]: Done.

File: haskeline.cabal (hunk 5 above)
Comment (RyanGlScott) [inline haskeline.cabal]: Sure. So when I originally started working on this PR, I noticed that there were two files with this exact boilerplate at the top:

```haskell
##if defined(i386_HOST_ARCH)
## define WINDOWS_CCONV stdcall
##elif defined(x86_64_HOST_ARCH)
## define WINDOWS_CCONV ccall
##else
## error Unknown mingw32 arch
##endif
```

And I was going to be adding a third file with the same boilerplate! So I moved this logic into `includes/windows_cconv.h`.

Note that `win_console.h` is put in an `install-includes` stanza of `haskeline.cabal`, which means that any library which transitively depends on `haskeline` can use `#include "win_console.h"` in their code. But `windows_cconv.h` is purely internal to `haskeline`, so I didn't put it in `install-includes`.

Sadly, for whatever reason `Cabal` doesn't put any files under an `includes` stanza into the tarball when you run `cabal sdist` (only the ones under `install-includes`). So to work around this, you have to list these files yourself in `extra-source-files`, which _are_ put into the tarball on a `cabal sdist`.

Alternatively, I could just remove `windows_cconv.h` entirely and go back to the way it was before, it you'd prefer that.

Comment (judah) [inline haskeline.cabal]: Thanks for the explanation.  My guess is it was a mistake to put `win_console.h` in `install-includes`, and it should also be listed in `extra-source-files` instead.  But OK to leave as-is for now.
